src/oic/oic/provider.py

```diff
@@ -2049,3 +1883,449 @@ def get_by_sub_and_(self, sub: str, key: str, val: Any) -> Optional[str]:
+        :param kwargs:
+        :return: Returns a dictionary with one key 'sjwt' and the value
+            being a signed JWT token with session information.
+        """
+        esr = EndSessionRequest().from_urlencoded(request)
+
+        logger.debug("End session request: {}", sanitize(esr.to_dict()))
+
+        if self.events:
+            self.events.store("protocol request", esr)
+
+        # 2 ways of find out client ID and user. Either through a cookie
+        # or using the id_token_hint. If I get information from both make sure they match
+        _, client_id, uid = self._get_uid_from_cookie(cookie)
+
+        sid = ""
+
+        if "id_token_hint" in esr:
+            id_token_hint = IdToken().from_jwt(
+                esr["id_token_hint"], keyjar=self.keyjar, verify=True
+            )
+            far_away = 86400 * 30  # 30 days
+
+            if client_id:
+                args = {"client_id": client_id}
+            else:
+                args = {}
+
+            try:
+                id_token_hint.verify(
+                    iss=self.baseurl, skew=far_away, nonce_storage_time=far_away, **args
+                )
+            except (VerificationError, NotForMe) as err:
+                logger.warning("Verification error on id_token_hint: {}".format(err))
+                return error_response("invalid_request", "Bad Id Token hint")
+
+            sub = id_token_hint["sub"]
+
+            if uid:
+                # verify that 'sub' are bound to 'uid'
+                if self.sdb.get_uid_by_sub(sub) != uid:
+                    return error_response("invalid_request", "Wrong user")
+            else:
+                uid = self.sdb.get_uid_by_sub(sub)
+
+            if not client_id:
+                if len(id_token_hint["aud"]) == 1:
+                    client_id = id_token_hint["aud"][0]
+                else:
+                    client_id = id_token_hint["azp"]
+
+            sids = session_get(self.sdb, "sub", sub)
+
+            for sid in sids:
+                if self.sdb[sid]["client_id"] == client_id:
+                    break
+
+        if not client_id:
+            return error_response("invalid_request", "Could not find client ID")
+        if client_id not in self.cdb:
+            return error_response("invalid_request", "Unknown client")
+
+        redirect_uri = None
+        if "post_logout_redirect_uri" in esr:
+            redirect_uri = self.verify_post_logout_redirect_uri(esr, client_id)
+            if not redirect_uri:
+                msg = "Post logout redirect URI verification failed!"
+                return error_response("invalid_request", msg)
+        else:  # If only one registered use that one
+            try:
+                _ruri = self.cdb[client_id]["post_logout_redirect_uris"]
+            except KeyError:
+                msg = "Missing post_logout_redirect_uri"
+                return error_response("invalid_request", msg)
+
+            if len(_ruri) == 1:
```

Comment:
`redirect_uri` is not defined for case where the `len(_ruri)` is different from `1`.

This is also not covered in tests. Either there are no logout URIs registered (and the SDB doesn't raise KeyError, which might be possible) or there are multiple.

Reply:
Done